File: dex_integrations/metis_jupiter.py
# =============================
# file: dex_integrations/metis_jupiter.py
# =============================
from __future__ import annotations
import os, typing as t
import httpx

Json = t.Dict[str, t.Any]

"""
Metis/Jupiter client with ordered fallback and proper bodies for /quote and /swap.
Env:
  METIS_BASE   = https://jupiter-swap-api.quiknode.pro/<YOUR_KEY>
  JUP_PRO      = https://api.jup.ag/swap/v1
  JUP_LITE     = https://lite-api.jup.ag/swap/v1
  PUBLIC_JUP   = https://www.jupiterapi.com
  JUP_API_KEY  = <optional, for Pro>
  JUP_TIMEOUT  = seconds (default 15)
"""

DEFAULT_BASES: t.List[str] = [
    os.getenv("METIS_BASE", "").rstrip("/"),
    os.getenv("JUP_PRO", "https://api.jup.ag/swap/v1").rstrip("/"),
    os.getenv("JUP_LITE", "https://lite-api.jup.ag/swap/v1").rstrip("/"),
    os.getenv("PUBLIC_JUP", "https://www.jupiterapi.com").rstrip("/"),
]
BASES = [b for b in DEFAULT_BASES if b]
TIMEOUT = float(os.getenv("JUP_TIMEOUT", "15"))


def _headers_for(base: str) -> Json:
    h = {"User-Agent": "metis-integration/1.0"}
    api_key = os.getenv("JUP_API_KEY")
    if api_key and ("api.jup.ag" in base):
        h["X-API-KEY"] = api_key
    return h


def _url(base: str, path: str) -> str:
    return f"{base}{path if path.startswith('/') else '/' + path}"


async def get_quote(
    input_mint: str,
    output_mint: str,
    amount: int,
    *,
    slippage_bps: int | None = None,
    swap_mode: str | None = None,  # "ExactIn" | "ExactOut"
    as_legacy: bool = False,
    dynamic_slippage: bool | None = None,
    extra: Json | None = None,
) -> Json:
    params: Json = {
        "inputMint": input_mint,
        "outputMint": output_mint,
        "amount": amount,
    }
    if slippage_bps is not None:
        params["slippageBps"] = slippage_bps
    if swap_mode:
        params["swapMode"] = swap_mode
    if as_legacy:
        params["asLegacyTransaction"] = True
    if dynamic_slippage is not None:
        params["dynamicSlippage"] = bool(dynamic_slippage)
    if extra:
        params.update(extra)

    last_err: Exception | None = None
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        for base in BASES:
            try:
                r = await client.get(_url(base, "/quote"), params=params, headers=_headers_for(base))
                if r.status_code == 200:
                    data = r.json()
                    if isinstance(data, dict) and (
                        data.get("routePlan") or data.get("outAmount") or data.get("otherAmountThreshold")
                    ):
                        return data
                last_err = RuntimeError(f"{base} /quote {r.status_code} {str(r.text)[:300]}")
            except Exception as e:  # network/timeout
                last_err = e
    raise last_err or RuntimeError("quote_failed")


async def build_swap_tx(
    quote_response: Json,
    user_public_key: str,
    *,
    wrap_and_unwrap_sol: bool = True,
    compute_unit_price_micro_lamports: int | None = None,
    as_legacy: bool = False,
    fee_account: str | None = None,
    destination_token_account: str | None = None,
    dynamic_cu_limit: bool = True,
    dynamic_slippage: bool | None = None,
    extra: Json | None = None,
) -> str:
    body: Json = {
        "userPublicKey": user_public_key,
        "quoteResponse": quote_response,
        "wrapAndUnwrapSol": wrap_and_unwrap_sol,
        "dynamicComputeUnitLimit": dynamic_cu_limit,
    }
    if compute_unit_price_micro_lamports is not None:
        body["computeUnitPriceMicroLamports"] = int(compute_unit_price_micro_lamports)
    if as_legacy:
        body["asLegacyTransaction"] = True
    if fee_account:
        body["feeAccount"] = fee_account
    if destination_token_account:
        body["destinationTokenAccount"] = destination_token_account
    if dynamic_slippage is not None:
        body["dynamicSlippage"] = bool(dynamic_slippage)
    if extra:
        body.update(extra)

    last_err: Exception | None = None
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        for base in BASES:
            try:
                r = await client.post(_url(base, "/swap"), json=body, headers=_headers_for(base))
                js = r.json() if r.headers.get("content-type", "").startswith("application/json") else {}
                if r.status_code == 200 and isinstance(js, dict) and js.get("swapTransaction"):
                    return js["swapTransaction"]
                if (r.status_code in (400, 422)) and not as_legacy:
                    body2 = dict(body)
                    body2["asLegacyTransaction"] = True
                    r2 = await client.post(_url(base, "/swap"), json=body2, headers=_headers_for(base))
                    js2 = r2.json() if r2.headers.get("content-type", "").startswith("application/json") else {}
                    if r2.status_code == 200 and js2.get("swapTransaction"):
                        return js2["swapTransaction"]
                last_err = RuntimeError(f"{base} /swap {r.status_code} {str(r.text)[:300]}")
            except Exception as e:
                last_err = e
    raise last_err or RuntimeError("swap_failed")


# --- Backward-compatible shims (keep old import sites working) ---
async def get_swap_route(input_mint: str, output_mint: str, amount: int):
    return await get_quote(input_mint, output_mint, amount)


async def get_swap_transaction(route_or_quote: Json, user_public_key: str):
    qr = route_or_quote
    if not isinstance(qr, dict) or not qr.get("inputMint"):
        raise ValueError("get_swap_transaction now requires the full quote object; call get_swap_route() first")
    return await build_swap_tx(qr, user_public_key, wrap_and_unwrap_sol=True)

File: trade-svc/src/metis-jupiter.ts
import axios, { AxiosError } from 'axios';
import https from 'node:https';

export type QuoteParams = {
  inputMint: string;
  outputMint: string;
  amountRaw: number; // base units
  slippageBps?: number;
  swapMode?: 'ExactIn' | 'ExactOut';
  asLegacyTransaction?: boolean;
  dynamicSlippage?: boolean;
  extra?: Record<string, any>;
};

export type SwapOpts = {
  userPublicKey: string;
  quote: any; // full quote object
  computeUnitPriceMicroLamports?: number;
  priorityFeeLamports?: number; // NEW: Direct lamports fee for Jupiter API
  asLegacyTransaction?: boolean;
  wrapAndUnwrapSol?: boolean;
  destinationTokenAccount?: string;
  feeAccount?: string;
  dynamicComputeUnitLimit?: boolean;
  dynamicSlippage?: boolean;
  extra?: Record<string, any>;
};

const httpsAgent = new https.Agent({ keepAlive: true, family: 4, maxSockets: 50 });
const UA = 'trade-svc/1.0';

const METIS_BASE = (process.env.METIS_BASE || '').replace(/\/$/, '');
const JUP_PRO    = (process.env.JUP_PRO    || 'https://api.jup.ag/swap/v1').replace(/\/$/, '');
const JUP_LITE   = (process.env.JUP_LITE   || 'https://lite-api.jup.ag/swap/v1').replace(/\/$/, '');
const PUBLIC_JUP = (process.env.PUBLIC_JUP || 'https://quote-api.jup.ag/v6').replace(/\/$/, '');

// Prioritize public API first, then others
const BASES = [PUBLIC_JUP, METIS_BASE, JUP_LITE, JUP_PRO].filter(Boolean);

function headersFor(base: string) {
  const h: Record<string, string> = { 'User-Agent': UA };
  const key = process.env.JUP_API_KEY;
  if (key && /api\.jup\.ag/.test(base)) h['X-API-KEY'] = key;
  return h;
}

function url(base: string, path: string) { return `${base}${path.startsWith('/') ? path : '/' + path}`; }

function mapAxiosErr(e: unknown): string {
  const ae = e as AxiosError;
  if (ae?.response) return `HTTP ${ae.response.status} ${String(JSON.stringify(ae.response.data)).slice(0, 300)}`;
  if (ae?.request)  return `network_error to ${(ae.config?.url)||'unknown'}: ${ae.code||ae.message}`;
  return (e as any)?.message || String(e);
}

export async function getQuote(p: QuoteParams): Promise<any> {
  const params = new URLSearchParams({
    inputMint: p.inputMint,
    outputMint: p.outputMint,
    amount: String(p.amountRaw),
  });
  if (p.slippageBps != null) params.set('slippageBps', String(p.slippageBps));
  if (p.swapMode) params.set('swapMode', p.swapMode);
  if (p.asLegacyTransaction) params.set('asLegacyTransaction', 'true');
  if (p.dynamicSlippage != null) params.set('dynamicSlippage', String(!!p.dynamicSlippage));
  if (p.extra) for (const [k, v] of Object.entries(p.extra)) params.set(k, String(v));

  let lastErr = '';
  for (const base of BASES) {
    try {
      const r = await axios.get(url(base, '/quote'), { httpsAgent, timeout: 12000, validateStatus: () => true, headers: headersFor(base), params });
      if (r.status === 200 && r.data && (r.data.routePlan || r.data.outAmount || r.data.otherAmountThreshold)) {
        return r.data; // full quote object
      }
      lastErr = `quote ${r.status} ${String(JSON.stringify(r.data)).slice(0, 300)}`;
    } catch (e) {
      lastErr = mapAxiosErr(e);
    }
  }
  throw new Error(lastErr || 'quote_failed');
}

export async function buildSwapTx(opts: SwapOpts): Promise<string> {
  const baseBody: Record<string, any> = {
    userPublicKey: opts.userPublicKey,
    quoteResponse: opts.quote,
    wrapAndUnwrapSol: opts.wrapAndUnwrapSol !== false,
    dynamicComputeUnitLimit: opts.dynamicComputeUnitLimit !== false,
  };
  
  // Priority fee handling - try multiple formats
  function addPriorityFeeToBody(body: any, base: string) {
    if (opts.priorityFeeLamports != null) {
      // For v6 API (quote-api.jup.ag) - use direct lamports format
      if (base.includes('quote-api.jup.ag')) {
        // Simple direct priority fee in lamports - this is the correct format for v6
        body.prioritizationFeeLamports = opts.priorityFeeLamports;
        console.log(`DEBUG Jupiter v6 API: Direct prioritizationFeeLamports = ${opts.priorityFeeLamports}`);
      }
      // For v1 APIs (api.jup.ag, lite-api.jup.ag) - use legacy format
      else {
        // Convert lamports to CU price (more aggressive conversion)
        const cuPrice = Math.round((opts.priorityFeeLamports * 10) / 1000);
        body.computeUnitPriceMicroLamports = cuPrice;
        console.log(`DEBUG Jupiter v1 API: computeUnitPriceMicroLamports = ${cuPrice} (from ${opts.priorityFeeLamports} lamports)`);
      }
    } else if (opts.computeUnitPriceMicroLamports != null) {
      body.computeUnitPriceMicroLamports = opts.computeUnitPriceMicroLamports;
      console.log(`DEBUG Jupiter API: computeUnitPriceMicroLamports = ${opts.computeUnitPriceMicroLamports}`);
    } else {
      console.log(`DEBUG Jupiter API: No priority fee parameters set`);
    }
  }
  
  if (opts.asLegacyTransaction) baseBody.asLegacyTransaction = true;
  if (opts.destinationTokenAccount) baseBody.destinationTokenAccount = opts.destinationTokenAccount;
  if (opts.feeAccount) baseBody.feeAccount = opts.feeAccount;
  if (opts.dynamicSlippage != null) baseBody.dynamicSlippage = !!opts.dynamicSlippage;
  if (opts.extra) Object.assign(baseBody, opts.extra);

  for (const base of BASES) {
    try {
      // Create body copy and add appropriate priority fee format
      const requestBody = { ...baseBody };
      addPriorityFeeToBody(requestBody, base);
      
      const r = await axios.post(url(base, '/swap'), requestBody, { 
        httpsAgent, 
        timeout: 15000, 
        validateStatus: () => true, 
        headers: { 'Content-Type': 'application/json', ...headersFor(base) } 
      });
      
      // Debug: Log Jupiter API response
      if (r.status === 200) {
        console.log(`✅ Jupiter API Success: ${base} responded with status 200`);
        if (r.data?.swapTransaction) {
          console.log(`✅ Got swapTransaction from Jupiter (length: ${r.data.swapTransaction.length})`);
          return r.data.swapTransaction as string;
        }
      } else {
        console.log(`❌ Jupiter API Error: ${base} responded with status ${r.status}`);
        console.log(`   Response: ${JSON.stringify(r.data).slice(0, 500)}`);
      }
      
      // Retry with legacy transaction if needed
      if ((r.status === 400 || r.status === 422) && !requestBody.asLegacyTransaction) {
        const retryBody = { ...requestBody, asLegacyTransaction: true };
        const r2 = await axios.post(url(base, '/swap'), retryBody, { 
          httpsAgent, 
          timeout: 15000, 
          validateStatus: () => true, 
          headers: { 'Content-Type': 'application/json', ...headersFor(base) } 
        });
        if (r2.status === 200 && r2.data?.swapTransaction) return r2.data.swapTransaction as string;
      }
    } catch (_) {
      // try next base
    }
  }
  throw new Error('swap_failed');
}
